### PR TITLE
Support non-square slab allocator pages in texture cache.

### PR DESCRIPTION
**Pull request by glennw**
Text runs are typically rectangles that are much longer in one
dimension than the other dimension (depending on horizontal or
vertical layout direction).

Storing rectangles of this shape in the texture cache was very
inefficient, since only square allocations are supported.

This patch adds support for texture cache regions to support
non-square slab sizes.

This is an efficiency win for any image that is rectangular,
and will become particularly important once we start caching
text runs in the texture cache.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2702)

### CODE REVIEW

**Comment by glennw** [comment]
r? @mrobinson or anyone

Pending try:
https://treeherder.mozilla.org/#/jobs?repo=try&revision=acae109e956c21ccc502c372970a72d8d3b66439

File: webrender/src/texture_cache.rs [hunk 1]
```diff
@@ -900,43 +901,49 @@ impl TextureCache {
...
-            _ => panic!("Invalid dimensions for cache!"),
+        let x_size = quantize_dimension(width);
+        let y_size = quantize_dimension(height);
+
+        assert!(x_size > 0 && x_size <= TEXTURE_REGION_DIMENSIONS);
+        assert!(y_size > 0 && y_size <= TEXTURE_REGION_DIMENSIONS);
+
+        let (width, height) = match (x_size, y_size) {
+            // Special cased rectangular slab pages.
+            (512, 512) => (512, 512),
```
Comment (mrobinson) [inline webrender/src/texture_cache.rs]: It seems like this line can be removed.

Comment (mrobinson) [inline webrender/src/texture_cache.rs]: It seems like this line can be removed, since it is just the square case?

Comment (glennw) [reply webrender/src/texture_cache.rs]: Good point! Fixed.

**Pull request merged**